src/gpu/nvidia/cudnn_binary.cpp

```diff
@@ -33,19 +34,22 @@ status_t cudnn_binary_t::execute(const exec_ctx_t &ctx) const {
             = utils::downcast<nvidia::sycl_cuda_stream_t *>(ctx.stream());
 
     return cuda_stream->interop_task([&](::sycl::handler &cgh) {
-        auto src_0_acc = CTX_IN_ACCESSOR(DNNL_ARG_SRC_0);
-        auto src_1_acc = CTX_IN_ACCESSOR(DNNL_ARG_SRC_1);
-        auto dst_acc = CTX_OUT_ACCESSOR(DNNL_ARG_DST);
+        auto *mem_src_0 = CTX_IN_MEMORY(DNNL_ARG_SRC_0);
+        auto *mem_src_1 = CTX_IN_MEMORY(DNNL_ARG_SRC_1);
+        auto *mem_dst = CTX_OUT_MEMORY(DNNL_ARG_DST);
+        auto src_0_acc = CTX_IN_OPTIONAL_ACCESSOR(DNNL_ARG_SRC_0, mem_src_0);
+        auto src_1_acc = CTX_IN_OPTIONAL_ACCESSOR(DNNL_ARG_SRC_1, mem_src_1);
+        auto dst_acc = CTX_OUT_OPTIONAL_ACCESSOR(DNNL_ARG_DST, mem_dst);
```

Comment:
I have a design idea that would allow to hide almost all implementation details and partially address the problem with sub-buffer (#1234).

The idea is to introduce an abstraction for a memory argument that will be used only by NVIDIA primitives. This abstraction can be generalized in the future but at this point it will only support CUDA backend. I suggest the following implementation of the abstraction:
```cpp
template <sycl::access_mode mode>
struct sycl_memory_arg_t {
    using acc_t = sycl::accessor<uint8_t, 1, mode>;

    // Constructor to instantiate the class for USM memory kind.
    sycl_memory_arg_t(void *usm) : usm_(usm) {}
    // Constructor to instantiate the class for buffer memory kind.
    sycl_memory_arg_t(const acc_t &acc) : usm_(nullptr), acc_(new acc_t(acc)) {}

    // This class will be responsible for extracting native memory object for CUDA backend.
    template <typename T = void>
    T *get_native_pointer(const sycl::interop_handle &ih) const {
        if(usm_) return reinterpret_cast<T *>(usm_);
        return reinterpret_cast<T *>(ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(*acc_));
    }
private:
    void *usm_;
    // SYCL 2020 accessors haven't been implemented yet hence using `std::shared_ptr`. 
    // This works only for `host_task`.
    std::shared_ptr<acc_t> acc_;
};
```
The primitives will work with the abstraction via a set of macros.
```cpp
#define CTX_IN_HOST_TASK_MEM_ARG(arg)
#define CTX_OUT_HOST_TASK_MEM_ARG(arg)
#define CTX_INOUT_HOST_TASK_MEM_ARG(arg)

// Implementation idea for `CTX_IN_HOST_TASK_MEM_ARG`.
// Buffer - access mode is automatically deduced. The abstraction can be extended to take an offset for sub-buffers.
sycl_memory_arg_t {buffer.get_access<sycl::access::mode::read>(cgh)}
// USM - requires specifying access mode for consistency.
sycl_memory_arg_t<sycl::access::mode::read> {usm_ptr}
```
With the aforementioned design almost all implementation details will be hidden and the code will be simplified significantly.
```cpp
return cuda_stream->interop_task([&](::sycl::handler &cgh) {
    auto src0 = CTX_IN_HOST_TASK_MEM_ARG(DNNL_ARG_SRC_0);
    auto src1 = CTX_IN_HOST_TASK_MEM_ARG(DNNL_ARG_SRC_0);
    auto dst = CTX_OUT_HOST_TASK_MEM_ARG(DNNL_ARG_DST);
    // ... //
    compat::host_task(cgh, [=](const compat::interop_handle &ih) {
        // ... //
        void *a = src0.get_native_pointer(ih);
        void *b = src1.get_native_pointer(ih);
        void *c = dst.get_native_pointer(ih);
        pd()->binary_impl_->execute(handle, a, b, c);
    });
});
```

Reply:
That is a good idea. I implemented it with some changes.